### PR TITLE
[PSST] Add parameter passing for policy script

### PR DESCRIPTION
**Pull request by vadimstruts**
Resolves https://github.com/brave/brave-browser/issues/48520

### CODE REVIEW

File: components/psst/browser/content/psst_script_utils.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#define BRAVE_COMPONENTS_PSST_BROWSER_CONTENT_PSST_SCRIPT_UTILS_H_
+
+#include <optional>
+#include <string>
+
+#include "base/values.h"
+
+namespace psst {
+
+std::string GetScriptWithParams(const std::string& script,
```
Comment (bridiver) [inline components/psst/browser/content/psst_script_utils.h]: why is this in a utils class instead of the web contents observer? It's not used anywhere else. If this is for testing only, this is not a great pattern to follow. Just put the method where it used and test it in `PsstTabWebContentsObserverUnitTest`. It really needs to be tested in the context of the actual script injection anyway. Just please create new tests specifically for testing params instead of adding it to existing tests in `PsstTabWebContentsObserverUnitTest`

Comment (vadimstruts) [reply components/psst/browser/content/psst_script_utils.h]: done

File: components/psst/browser/content/psst_tab_web_contents_observer.cc [hunk 2]
```diff
@@ -30,6 +33,23 @@ struct PsstNavigationData : public base::SupportsUserData::Data {
   int id;
 };
 
+std::string GetScriptWithParams(const std::string& script, base::Value params) {
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer.cc]: nit: `MaybeAddParamsToScript` would be better here I think. `Maybe` is normally used as a prefix for methods that do something conditionally

File: components/psst/browser/content/psst_tab_web_contents_observer.cc [hunk 3]
```diff
@@ -30,6 +33,23 @@ struct PsstNavigationData : public base::SupportsUserData::Data {
...
+  const auto* params_dict = params.GetIfDict();
+  if (!params_dict || params_dict->empty()) {
+    return script;
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer.cc]: @ShivanKaul this would indicate the script itself is broken somehow? Do we want to be notified about this? If so, we could set crash keys and DumpWithoutCrashing here

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 2 above)
Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: some comments here describing the format of the params would be helpful as well

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 3 above)
Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: actually we are also checking here https://github.com/brave/brave-core/pull/30716/files#diff-4ff1e66bb3980dd4de7badd59e7641673c2bb28412d1b093746738fc03c8efa3R152
so at this point it must be a dict and only an empty dict would indicate a problem?

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 2 above)
Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: You already know this is a Dict because you checked here https://github.com/brave/brave-core/pull/30716/files#diff-4ff1e66bb3980dd4de7badd59e7641673c2bb28412d1b093746738fc03c8efa3R152 so you just pass a Dict instead of Value

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc [hunk 4]
```diff
@@ -460,6 +500,39 @@ TEST_F(PsstTabWebContentsObserverUnitTest,
   policy_script_insert_loop.Run();
 }
 
+TEST_F(PsstTabWebContentsObserverUnitTest,
+       UserScriptReturnsWrongDictValuesNoPolicyScriptParams) {
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: wrong in what way?

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc [hunk 5]
```diff
@@ -440,6 +442,44 @@ TEST_F(PsstTabWebContentsObserverUnitTest,
...
+                                                             url);
+  observer.Wait();
+
+  check_loop.Run();
+  user_script_insert_loop.Run();
+  policy_script_insert_loop.Run();
+}
+
+TEST_F(PsstTabWebContentsObserverUnitTest,
+       UserScriptReturnsEmptyDictNoPolicyScriptParams) {
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: Seems to be missing a test for has policy and empty dict. I believe this might indicate a misbehaving script as mentioned above? Not sure, but either way it's covered in the code and should have a test as well

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc [hunk 6]
```diff
@@ -440,6 +442,44 @@ TEST_F(PsstTabWebContentsObserverUnitTest,
           &check_loop, CreateMatchedRule(user_script, policy_script)));
   base::RunLoop user_script_insert_loop;
   base::RunLoop policy_script_insert_loop;
+  auto dict = base::Value(base::Value::Dict().Set("prop", "value"));
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: It would be more useful to name this `script_params` or similar

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc [hunk 7]
```diff
@@ -440,6 +442,44 @@ TEST_F(PsstTabWebContentsObserverUnitTest,
...
+  auto value = base::Value();
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: policy_script_return_value?

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 3 above)
Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: maybe we don't care if this happens or not (the empty dict), but we are duplicating checks and passing a Value that we already know is a Dict

Comment (ShivanKaul) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: Makes sense to me that we should only pass in a `base::Dict` to `GetScriptWithParams`.

Comment (ShivanKaul) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: > @ShivanKaul this would indicate the script itself is broken somehow?

@vadimstruts is looking into this.

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 2 above)
Comment (ShivanKaul) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: > MaybeAddParamsToScript

Yeah this is a better name, it took me a while to figure out what this function was actually doing.

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc (hunk 4 above)
Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: renamed it to `UserScriptReturnsUnsupportedDictNoPolicyScriptParams` as in case of  dictionary with a blob storage value, we can't serialize.
Added a comment in the code

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 3 above)
Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: Added crash keys and DumpWithoutCrashing for empty dictionary and failed serialization

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc (hunk 5 above)
Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: do you mean the case when we have empty script parameters dict, and as result execute policy script without any parameters?
If yes, Seems It exists: UserScriptReturnsEmptyDictNoPolicyScriptParams

File: components/psst/browser/content/psst_tab_web_contents_observer.cc (hunk 3 above)
Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: Perhaps we need to decompose the following construction https://github.com/brave/brave-core/blob/95fb376abd4374f983e15ef103ae8ec8bb387106/components/psst/browser/content/psst_tab_web_contents_observer.cc#L175
to have the `!user_script_result.is_dict()` as the separate `if` block and have crash keys and DumpWithoutCrashing there?

File: components/psst/browser/content/psst_tab_web_contents_observer.cc [hunk 8]
```diff
@@ -30,6 +35,44 @@ struct PsstNavigationData : public base::SupportsUserData::Data {
...
+  if (params_dict.empty()) {
+    SCOPED_CRASH_KEY_STRING64("Psst", "params_dict", "empty");
+    base::debug::DumpWithoutCrashing();
+    return script;
+  }
+
+  std::optional<std::string> params_json = base::WriteJsonWithOptions(
+      params_dict, base::JSONWriter::OPTIONS_PRETTY_PRINT);
+  if (!params_json) {
+    SCOPED_CRASH_KEY_STRING64("Psst", "params_dict", "could not be serialized");
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer.cc]: this doesn't seem very useful? I think we'd want to know which script, but I guess we have the url already so maybe we don't need anything here?

Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: we definitely don't need to know things like `could not be serialized` because we know exactly where this happened since we get the stack trace, we only need context we can't get from the stack trace

Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc]: To add more context, added psst script version and rule name (related with url) crash keys instead

File: components/psst/browser/content/psst_tab_web_contents_observer.cc [hunk 9]
```diff
@@ -30,6 +35,47 @@ struct PsstNavigationData : public base::SupportsUserData::Data {
...
+    SCOPED_CRASH_KEY_STRING64("Psst", "rule_name", rule->name());
+    SCOPED_CRASH_KEY_NUMBER("Psst", "rule_version", rule->version());
+    base::debug::DumpWithoutCrashing();
+    return rule->policy_script();
+  }
+
+  std::optional<std::string> params_json = base::WriteJsonWithOptions(
+      params_dict, base::JSONWriter::OPTIONS_PRETTY_PRINT);
+  if (!params_json) {
+    SCOPED_CRASH_KEY_STRING64("Psst", "rule_name", rule->name());
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer.cc:56]: why are you duplicating these? Just set them at a higher scope

Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer.cc:56]: done

File: components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc (hunk 5 above)
Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: I guess to clarify there are two cases that need to be tested
1. dict is empty - this is an error condition
2. there is no dict - this is normal

Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: looks like both are covered

Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: actually I'm confused about what this test is checking vs https://github.com/brave/brave-core/pull/30716/files#diff-9e04e8e6680ea5deda467bc022af53bd84f5abe0dafd445fb14a1408b34fc686R548

Comment (bridiver) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: I think you needs some comments explaining these because I'm not sure what `NoPolicyScriptParams` is supposed to mean here. Also if the dict is empty, there is a policy script, right? Because the dict is not supposed to be empty

Comment (vadimstruts) [reply components/psst/browser/content/psst_tab_web_contents_observer_unittest.cc]: If `script_params` is an empty dictionary, It means that that after execution of the user script returned empty dictionary, and it is OK for policy script execution.
In this test we have next line: https://github.com/brave/brave-core/pull/30716/files#diff-9e04e8e6680ea5deda467bc022af53bd84f5abe0dafd445fb14a1408b34fc686R541
So the policy script is empty, and as result not executed

File: components/psst/browser/content/psst_tab_web_contents_observer.cc [hunk 10]
```diff
@@ -30,6 +35,45 @@ struct PsstNavigationData : public base::SupportsUserData::Data {
...
+// function returns: const params = {
+//    "tasks": [ {
+//       "description": "Ads Preferences",
+//       "url": "https://a.test/settings/ads_preferences"
+//    } ]
+// };
+// <policy script, which uses parameters to apply PSST settings selected by the
+// user>;
+//
+// @param rule Psst rule used for current URL to apply privacy settings.
```
Comment (bridiver) [inline components/psst/browser/content/psst_tab_web_contents_observer.cc]: We don't use doxygen so please don't use doxygen tags. We don't need a description of the params/return values

**Review by bridiver — APPROVED** [review]

File: components/psst/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      {
+        url:<setting url, MUST BE UNIQUE>,
+        description:<setting description>,
+      },
+       .... 
+   }
+ }
+```
+
+#### privacy.js
```
Comment (ShivanKaul) [inline components/psst/README.md:85]: This should be `policy.js`.
